### PR TITLE
Downgrade sentinel

### PR DESCRIPTION
**Pull request by tomato42**
### Description

Fix handling of downgrade sentinel when highest version is negotiated

### Motivation and Context

https://gitlab.com/gnutls/gnutls/issues/689

### Checklist

- [x] I have read the [CONTRIBUTING.md](https://github.com/tomato42/tlsfuzzer/blob/master/CONTRIBUTING.md) document and my PR follows [change requirements](https://github.com/tomato42/tlsfuzzer/blob/master/CONTRIBUTING.md#change-requirements) therein
- [x] the changes are also reflected in documentation and code comments
- [x] all new and existing tests pass (see Travis CI results)
- [x] [test script checklist](https://github.com/tomato42/tlsfuzzer/wiki/Test-script-checklist) was followed for new scripts
- [x] new test script added to `tlslite-ng.json` and `tlslite-ng-random-subset.json`
- [x] new and modified scripts were ran against popular TLS implementations:
  - [x] OpenSSL
  - [x] NSS
  - [x] GnuTLS [GnuTLS#689](https://gitlab.com/gnutls/gnutls/issues/689)
- [x] required version of tlslite-ng updated in requirements.txt and README.md

---
This change is [Reviewable](https://reviewable.io/reviews/tomato42/tlsfuzzer/507)

### CODE REVIEW

**Comment by tomato42** [comment]
@simo5 could you review this, since you're the original author of this code?

File: tlsfuzzer/expect.py [hunk 1]
```diff
@@ -372,10 +372,32 @@ def __init__(self, extensions=None, version=None, resume=False,
...
+        send no extensions. Order does not matter, but all extensions present
+        and only extensions present in the list must be sent by server. None
+        as the value of the relevant extension type can be used to select
+        autohandler for a given extension type.
+
+        @param tuple version: the literal version in the Server Hello message
+        (needs to be (3, 3) for TLS 1.3, use extensions to expect TLS 1.3
+        negotiation)
+
+        @param typle server_max_protocol: the higher protocol version supported
```
Comment (simo5) [inline tlsfuzzer/expect.py]: typle -> tuple

File: tlsfuzzer/expect.py [hunk 2]
```diff
@@ -372,10 +372,32 @@ def __init__(self, extensions=None, version=None, resume=False,
...
+        by server. Used for testing downgrade signalling of servers.
```
Comment (simo5) [inline tlsfuzzer/expect.py]: signalling -> signaling

File: tests/test_tlsfuzzer_expect.py [hunk 3]
```diff
@@ -1538,9 +1632,12 @@ def test_process_with_tls_1_1_wrong_downgrade_protection(self):
 
         self.assertTrue(exp.is_match(msg))
 
-        with self.assertRaises(AssertionError):
+        with self.assertRaises(AssertionError) as e:
             exp.process(state, msg)
 
+        self.assertIn("downgrade protection sentinel but shouldn",
```
Comment (simo5) [inline tests/test_tlsfuzzer_expect.py]: incorrectly indented ?

File: tests/test_tlsfuzzer_expect.py [hunk 4]
```diff
@@ -1432,9 +1523,12 @@ def test_process_with_tls_1_2_wrong_downgrade_protection(self):
...
```
Comment (simo5) [inline tests/test_tlsfuzzer_expect.py]: how does e get here ? shouldn't this be indented in the with statement above ?

File: tests/test_tlsfuzzer_expect.py [hunk 5]
```diff
@@ -1285,52 +1285,140 @@ def test_process_with_tls13_unallowed_extension(self):
...
+                                   cipher_suite=
+                                   CipherSuite.TLS_AES_128_GCM_SHA256,
+                                   extensions=ext)
+
+        self.assertTrue(exp.is_match(msg))
+
+        with self.assertRaises(AssertionError) as e:
+            exp.process(state, msg)
+
+        self.assertIn("downgrade protection sentinel but shouldn",
```
Comment (simo5) [inline tests/test_tlsfuzzer_expect.py]: clarify message ?
also shouldn -> shouldn't ?

File: tests/test_tlsfuzzer_expect.py [hunk 6]
```diff
@@ -1285,52 +1285,140 @@ def test_process_with_tls13_unallowed_extension(self):
...
+                                   CipherSuite.TLS_AES_128_GCM_SHA256,
+                                   extensions=ext)
 
         self.assertTrue(exp.is_match(msg))
 
-        with self.assertRaises(AssertionError):
+        with self.assertRaises(AssertionError) as e:
             exp.process(state, msg)
 
+        self.assertIn("downgrade protection sentinel but shouldn",
```
Comment (simo5) [inline tests/test_tlsfuzzer_expect.py]: shouldn -> shouldn't
mesage is unclear anyhow

File: tests/test_tlsfuzzer_expect.py [hunk 7]
```diff
@@ -1285,52 +1285,140 @@ def test_process_with_tls13_unallowed_extension(self):
...
 
         self.assertTrue(exp.is_match(msg))
 
         exp.process(state, msg)
 
-    def test_process_with_tls_1_3_wrong_downgrade_protection(self):
-        exp = ExpectServerHello(extensions={ExtensionType.renegotiation_info:
-                                            None},
-                                version=(3, 4), server_max_protocol=(3, 4))
+    def test_process_with_default_and_tls_1_3_with_1_2_downgrade_sentinel(self):
```
Comment (simo5) [inline tests/test_tlsfuzzer_expect.py]: name of test is confusing to me

File: tests/test_tlsfuzzer_expect.py [hunk 8]
```diff
@@ -1285,52 +1285,140 @@ def test_process_with_tls13_unallowed_extension(self):
...
         state = ConnectionState()
         client_hello = ClientHello()
         ciph = CipherSuite.TLS_EMPTY_RENEGOTIATION_INFO_SCSV
         client_hello.cipher_suites = [4, ciph]
         state.handshake_messages.append(client_hello)
         state.msg_sock = mock.MagicMock()
+        ext = []
+        ext.append(SrvSupportedVersionsExtension().create((3, 4)))
 
         msg = ServerHello().create(version=(3, 4),
```
Comment (simo5) [inline tests/test_tlsfuzzer_expect.py]: is version here ok ?

File: scripts/test-downgrade-protection.py [hunk 9]
```diff
@@ -86,6 +86,10 @@ def main():
     else:
         run_only = None
 
+    if srv_max_prot is None:
+        raise ValueError("Highest protocol version supported by server unset,"
+                         " use --server-max-protocol to set it")
+
```
Comment (simo5) [inline scripts/test-downgrade-protection.py]: why this change ?
The semantic here was that an unspecified max prot means the maximum we support (TLS 1.3)

File: tests/test_tlsfuzzer_expect.py (hunk 3 above)
Comment (tomato42) [reply tests/test_tlsfuzzer_expect.py]: no? assertRaises does not abort execution, exactly so that raised exception can be inspected, as we do here

File: tests/test_tlsfuzzer_expect.py (hunk 5 above)
Comment (tomato42) [reply tests/test_tlsfuzzer_expect.py]: it's just check that the raised exception is the one we expect, not just any AssertionError, but yeah the `'t` could be added - will do

File: scripts/test-downgrade-protection.py (hunk 9 above)
Comment (tomato42) [reply scripts/test-downgrade-protection.py]: it mostly is a leftover when I was trying to figure out why the gnutls bug wasn't being detected, it indeed is unnecessary now, I'll drop that patch

File: tlsfuzzer/expect.py (hunk 1 above)
Comment (tomato42) [reply tlsfuzzer/expect.py]: done

File: tlsfuzzer/expect.py (hunk 2 above)
Comment (tomato42) [reply tlsfuzzer/expect.py]: done

File: tests/test_tlsfuzzer_expect.py (hunk 4 above)
Comment (tomato42) [reply tests/test_tlsfuzzer_expect.py]: just as above - assertRaises catches the exception

File: tests/test_tlsfuzzer_expect.py (hunk 5 above)
Comment (tomato42) [reply tests/test_tlsfuzzer_expect.py]: the exception and message in question:
https://github.com/tomato42/tlsfuzzer/blob/c28cdd361f3255f1a76d652c79e0af04c4541103/tlsfuzzer/expect.py#L663-L665

File: tests/test_tlsfuzzer_expect.py (hunk 7 above)
Comment (tomato42) [reply tests/test_tlsfuzzer_expect.py]: changed it to `test_process_with_default_settings_and_tls_1_3_reply_with_1_2_downgrade_sentinel` and extended the comment below, does that make it clearer? if not any other suggestion how to make it clear?

File: tests/test_tlsfuzzer_expect.py (hunk 8 above)
Comment (tomato42) [reply tests/test_tlsfuzzer_expect.py]: yes, this is exactly to verify that the server doesn't advertise the negotiated version incorrectly

File: tests/test_tlsfuzzer_expect.py [hunk 10]
```diff
@@ -1285,52 +1285,142 @@ def test_process_with_tls13_unallowed_extension(self):
...
 
         self.assertTrue(exp.is_match(msg))
 
         exp.process(state, msg)
 
-    def test_process_with_tls_1_3_wrong_downgrade_protection(self):
-        exp = ExpectServerHello(extensions={ExtensionType.renegotiation_info:
-                                            None},
-                                version=(3, 4), server_max_protocol=(3, 4))
+    def test_process_with_default_settings_and_tls_1_3_reply_with_1_2_downgrade_sentinel(self):
```
Comment (simo5) [inline tests/test_tlsfuzzer_expect.py]: I think I prefer something like test_process_with_incorrect_tls_1_2_downgrade_sentinel()

File: tests/test_tlsfuzzer_expect.py [hunk 11]
```diff
@@ -1285,52 +1285,142 @@ def test_process_with_tls13_unallowed_extension(self):
...
         self.assertTrue(exp.is_match(msg))
 
-        with self.assertRaises(AssertionError):
+        with self.assertRaises(AssertionError) as e:
             exp.process(state, msg)
 
+        self.assertIn("downgrade protection sentinel but shouldn't",
+                      str(e.exception))
+
+    def test_process_with_default_settings_and_tls_1_3_reply_with_1_1_downgrade_sentinel(self):
```
Comment (simo5) [inline tests/test_tlsfuzzer_expect.py]: as above I would shorten this to test_process_with_incorrect_tls_1_1_downgrade_sentinel()

File: tests/test_tlsfuzzer_expect.py (hunk 10 above)
Comment (tomato42) [reply tests/test_tlsfuzzer_expect.py]: the thing is that you can have two incorrect downgrade sentinels: 
 1. when a TLS 1.3 Server Hello includes it
 2. when the server actually implements TLS 1.3 but downgraded to TLS 1.2 and included the TLS 1.2 sentinel

this is the former
